executor/src/witgen/jit/interpreter.rs

```diff
@@ -0,0 +1,391 @@
+                            MachineCallArgument::Unknown(v) => (Some(v), Default::default()),
+                            MachineCallArgument::Known(v) => (
+                                None,
+                                self.evaluate_expression(&mut eval_stack, &vars, &v.to_rpn()),
+                            ),
+                        })
+                        .collect();
+
+                    // call machine
+                    let mut args = arguments
+                        .iter()
+                        .zip(arg_values.iter_mut())
+                        .map(|(a, v)| match a {
+                            MachineCallArgument::Unknown(_) => LookupCell::Output(&mut v.1),
+                            MachineCallArgument::Known(_) => LookupCell::Input(&v.1),
+                        })
+                        .collect::<Vec<_>>();
+
+                    mutable_state.call_direct(*id, &mut args[..]).unwrap();
+
+                    // write output to variables
+                    let mut var_idx = 0;
+                    args.into_iter().for_each(|arg| {
+                        if let LookupCell::Output(v) = arg {
+                            assert!(vars[result_vars[var_idx]].replace(*v).is_none());
+                            var_idx += 1;
+                        }
+                    });
+                }
+                InterpreterAction::Assertion(e1, e2, expected_equal) => {
+                    let lhs_value = self.evaluate_expression(&mut eval_stack, &vars, e1);
+                    let rhs_value = self.evaluate_expression(&mut eval_stack, &vars, e2);
+                    if *expected_equal {
+                        assert_eq!(lhs_value, rhs_value, "Assertion failed");
+                    } else {
+                        assert_ne!(lhs_value, rhs_value, "Assertion failed");
+                    }
+                }
+            }
+        }
+    }
+
+    // evaluates an expression using the provided variable values
+    fn evaluate_expression(
+        &self,
+        stack: &mut Vec<T>,
+        vars: &[Option<T>],
+        expr: &RPNExpression<T, usize>,
+    ) -> T {
+        for elem in expr.elems.iter() {
+            match elem {
+                RPNExpressionElem::Concrete(v) => stack.push(*v),
+                RPNExpressionElem::Symbol(idx) => stack.push(vars[*idx].unwrap()),
+                RPNExpressionElem::BinaryOperation(op) => {
+                    let right = stack.pop().unwrap();
+                    let left = stack.pop().unwrap();
+                    let result = match op {
+                        BinaryOperator::Add => left + right,
+                        BinaryOperator::Sub => left - right,
+                        BinaryOperator::Mul => left * right,
+                        BinaryOperator::Div => left / right,
+                        BinaryOperator::IntegerDiv => T::from(
+                            left.to_integer().try_into_u64().unwrap()
```

Comment:
I think we have to turn this into arbitrary integer. Or (maybe better) we require division on `FieldElement::Integer`.

Reply:
I'm not sure where's the right place to put the `Div` bound on `FieldElement::Integer`, if that's what you mean